integration/test/Test/EnterpriseLogin.hs

```diff
@@ -0,0 +1,291 @@
+    resp.status `shouldMatchInt` 200
+    resp.json %. "domain" `shouldMatch` domain
+    resp.json %. "domain_redirect" `shouldMatch` "locked"
+    resp.json %. "team_invite" `shouldMatch` "allowed"
+  -- remove from deny-list
+  assertStatus 204 =<< domainRegistrationUnlock OwnDomain domain
+  -- check that it got removed
+  bindResponse (getDomainRegistration OwnDomain domain) $ \resp -> do
+    resp.status `shouldMatchInt` 200
+    resp.json %. "domain" `shouldMatch` domain
+    resp.json %. "domain_redirect" `shouldMatch` "none"
+    resp.json %. "team_invite" `shouldMatch` "allowed"
+
+testDomainRegistrationLockPreviousValueOverwritten :: App ()
+testDomainRegistrationLockPreviousValueOverwritten = do
+  domain <- randomDomain
+  -- pre-authorize
+  assertStatus 204 =<< domainRegistrationPreAuthorize OwnDomain domain
+  bindResponse (getDomainRegistration OwnDomain domain) $ \resp -> do
+    resp.status `shouldMatchInt` 200
+    resp.json %. "domain_redirect" `shouldMatch` "pre-authorized"
+  -- lock
+  assertStatus 204 =<< domainRegistrationLock OwnDomain domain
+  -- check that it got overwritten
+  bindResponse (getDomainRegistration OwnDomain domain) $ \resp -> do
+    resp.status `shouldMatchInt` 200
+    resp.json %. "domain_redirect" `shouldMatch` "locked"
+
+testDomainRegistrationUnlockErrorIfNotLocked :: App ()
+testDomainRegistrationUnlockErrorIfNotLocked = do
+  domain <- randomDomain
+  -- pre-authorize
+  assertStatus 204 =<< domainRegistrationPreAuthorize OwnDomain domain
+  bindResponse (getDomainRegistration OwnDomain domain) $ \resp -> do
+    resp.status `shouldMatchInt` 200
+    resp.json %. "domain_redirect" `shouldMatch` "pre-authorized"
+  -- attempt to unlock should fail
+  bindResponse (domainRegistrationUnlock OwnDomain domain) $ \resp -> do
+    resp.status `shouldMatchInt` 400
+    resp.json %. "label" `shouldMatch` "unlock-error"
+
+testDomainRegistrationPreAuthorize :: App ()
+testDomainRegistrationPreAuthorize = do
+  domain <- randomDomain
+  -- it should not yet exist
+  assertStatus 404 =<< getDomainRegistration OwnDomain domain
+  -- pre-authorize
+  assertStatus 204 =<< domainRegistrationPreAuthorize OwnDomain domain
+  -- idempotent
+  assertStatus 204 =<< domainRegistrationPreAuthorize OwnDomain domain
+  -- it got created
+  bindResponse (getDomainRegistration OwnDomain domain) $ \resp -> do
+    resp.status `shouldMatchInt` 200
+    resp.json %. "domain" `shouldMatch` domain
+    resp.json %. "domain_redirect" `shouldMatch` "pre-authorized"
+    resp.json %. "team_invite" `shouldMatch` "allowed"
+
+testDomainRegistrationPreAuthorizeFailsIfLocked :: App ()
+testDomainRegistrationPreAuthorizeFailsIfLocked = do
+  domain <- randomDomain
+  -- add to deny-list
+  assertStatus 204 =<< domainRegistrationLock OwnDomain domain
+  -- pre-authorize
+  bindResponse (domainRegistrationPreAuthorize OwnDomain domain) $ \resp -> do
+    resp.status `shouldMatchInt` 400
```

Comment:
400 is "something was wrong with your request", but this is more of a "the server state does not allow this request at this moment".  "409 conflict"?